File: mvvm/src/main/java/com/lee/mvvm/base/view/widge/rv/EmptyRecyclerView.java
package com.lee.mvvm.base.view.widge.rv;

import android.content.Context;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

public class EmptyRecyclerView extends RecyclerView {

    private View mEmptyView;

    private AdapterDataObserver mObserver = new AdapterDataObserver() {
        @Override
        public void onChanged() {
            Adapter adapter = getAdapter();
            if (adapter == null || mEmptyView == null) {
                return;
            }
            if (adapter.getItemCount() == 0) {
                mEmptyView.setVisibility(VISIBLE);
                EmptyRecyclerView.this.setVisibility(GONE);
            } else {
                mEmptyView.setVisibility(GONE);
                EmptyRecyclerView.this.setVisibility(VISIBLE);
            }
        }

        public void onItemRangeChanged(int positionStart, int itemCount) {
            onChanged();
        }

        public void onItemRangeMoved(int fromPosition, int toPosition, int itemCount) {
            onChanged();
        }

        public void onItemRangeRemoved(int positionStart, int itemCount) {
            onChanged();
        }

        public void onItemRangeInserted(int positionStart, int itemCount) {
            onChanged();
        }

        public void onItemRangeChanged(int positionStart, int itemCount, Object payload) {
            onChanged();
        }
    };

    public EmptyRecyclerView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    public void setEmptyView(@NonNull View view) {
        this.mEmptyView = view;
        ((ViewGroup) this.getParent()).addView(mEmptyView);
    }

    public void setEmptyView(@LayoutRes int layoutId) {
        LayoutInflater inflater = LayoutInflater.from(getContext());
        this.mEmptyView = inflater.inflate(layoutId, (ViewGroup) getParent(), false);
        ((ViewGroup) this.getParent()).addView(mEmptyView);
    }

    public void setAdapter(RecyclerView.Adapter adapter) {
        super.setAdapter(adapter);
        adapter.registerAdapterDataObserver(mObserver);
        mObserver.onChanged();
    }
}

File: mvvm/src/main/java/com/lee/mvvm/base/view/widge/rv/EmptyRecyclerView.kt
package com.lee.mvvm.base.view.widge.rv

import android.content.Context
import androidx.annotation.LayoutRes
import androidx.recyclerview.widget.RecyclerView
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

class EmptyRecyclerView(context: Context, attrs: AttributeSet?) : RecyclerView(context, attrs) {

    private var mEmptyView: View? = null

    private val mObserver = object : RecyclerView.AdapterDataObserver() {
        override fun onChanged() {
            val adapter = adapter
            if (adapter == null || mEmptyView == null) {
                return
            }
            if (adapter.itemCount == 0) {
                mEmptyView!!.visibility = View.VISIBLE
                this@EmptyRecyclerView.visibility = View.GONE
            } else {
                mEmptyView!!.visibility = View.GONE
                this@EmptyRecyclerView.visibility = View.VISIBLE
            }
        }

        override fun onItemRangeChanged(positionStart: Int, itemCount: Int) {
            onChanged()
        }

        override fun onItemRangeMoved(fromPosition: Int, toPosition: Int, itemCount: Int) {
            onChanged()
        }

        override fun onItemRangeRemoved(positionStart: Int, itemCount: Int) {
            onChanged()
        }

        override fun onItemRangeInserted(positionStart: Int, itemCount: Int) {
            onChanged()
        }

        override fun onItemRangeChanged(positionStart: Int, itemCount: Int, payload: Any?) {
            onChanged()
        }
    }

    fun setEmptyView(view: View) {
        this.mEmptyView = view
        (this.parent as ViewGroup).addView(mEmptyView, (this.parent as ViewGroup).indexOfChild(this))
    }

    fun setEmptyView(@LayoutRes layoutId: Int) {
        val inflater = LayoutInflater.from(context)
        this.mEmptyView = inflater.inflate(layoutId, parent as ViewGroup, false)
        (this.parent as ViewGroup).addView(mEmptyView, (this.parent as ViewGroup).indexOfChild(this))
    }

    override fun setAdapter(adapter: RecyclerView.Adapter<*>?) {
        super.setAdapter(adapter)
        adapter!!.registerAdapterDataObserver(mObserver)
        mObserver.onChanged()
    }
}